src/io/windows.zig

```diff
@@ -0,0 +1,1132 @@
+            context: Context,
+            completion: *Completion,
+            result: TimeoutError!void,
+        ) void,
+        completion: *Completion,
+        nanoseconds: u63,
+    ) void {
+        self.submit(
+            context,
+            callback,
+            completion,
+            .timeout,
+            .{ .deadline = self.timer.monotonic() + nanoseconds },
+            struct {
+                fn do_operation(ctx: Completion.Context, op: anytype) TimeoutError!void {
+                    _ = ctx;
+                    _ = op;
+                    return;
+                }
+            },
+        );
+    }
+
+    pub const INVALID_SOCKET = os.windows.ws2_32.INVALID_SOCKET;
+
+    pub fn open_socket(self: *IO, family: u32, sock_type: u32, protocol: u32) !os.socket_t {
+        var flags: os.windows.DWORD = 0;
+        flags |= os.windows.ws2_32.WSA_FLAG_OVERLAPPED;
+        flags |= os.windows.ws2_32.WSA_FLAG_NO_HANDLE_INHERIT;
+
+        const socket = try os.windows.WSASocketW(
+            @bitCast(i32, family),
+            @bitCast(i32, sock_type),
+            @bitCast(i32, protocol),
+            null,
+            0,
+            flags,
+        );
+        errdefer os.closeSocket(socket);
+
+        const socket_iocp = try os.windows.CreateIoCompletionPort(socket, self.iocp, 0, 0);
+        assert(socket_iocp == self.iocp);
+
+        // Ensure that synchronous IO completion doesn't queue an unneeded overlapped
+        // and that the event for the socket (WaitForSingleObject) doesn't need to be set.
+        var mode: os.windows.BYTE = 0;
+        mode |= os.windows.FILE_SKIP_COMPLETION_PORT_ON_SUCCESS;
+        mode |= os.windows.FILE_SKIP_SET_EVENT_ON_HANDLE; 
+
+        const handle = @ptrCast(os.windows.HANDLE, socket);
+        try os.windows.SetFileCompletionNotificationModes(handle, mode);
+
+        return socket;
+    }
+
+    pub fn open_dir(dir_path: [:0]const u8) !os.fd_t {
+        const dir = try std.fs.cwd().openDirZ(dir_path, .{});
+        return dir.fd;
+    }
+
+    pub fn open_file(
+        self: *IO,
+        dir_handle: os.fd_t,
+        relative_path: [:0]const u8,
+        size: u64,
+        must_create: bool,
+    ) !os.fd_t {
+        const path_w = try os.windows.sliceToPrefixedFileW(relative_path);
+
+        // FILE_CREATE = O_CREAT | O_EXCL
+        var creation_disposition: os.windows.DWORD = 0;
+        if (must_create) {
+            log.info("creating \"{s}\"...", .{relative_path});
+            creation_disposition = os.windows.FILE_CREATE;
+        } else {
+            log.info("opening \"{s}\"...", .{relative_path});
+            creation_disposition = os.windows.OPEN_EXISTING;
+        }
+
+        // O_EXCL
+        var shared_mode: os.windows.DWORD = 0;
+
+        // O_RDWR
+        var access_mask: os.windows.DWORD = 0;
+        access_mask |= os.windows.GENERIC_READ;
+        access_mask |= os.windows.GENERIC_WRITE;
+
+        // O_DIRECT
```

Comment:
```suggestion
        // O_DIRECT | O_DSYNC
```

Reply:
There's a subtle difference between both of these. They're both necessary for correctness, but neither are sufficient.

`O_DIRECT` means we bypass the kernel's page cache on the read and write path:
* Bypassing the page cache on the read path is critical for correctness, and means we see what's actually on the disk when recovering at startup, so we can evaluate how much data is truly durable, before we make any external guarantees about durability to the rest of the cluster. It's quite possible for the kernel to get out of sync with the disk and to have a clean page in the page cache that's not actually on disk, which the kernel will never try to write again either.
* Bypassing the page cache on the write path is great for performance, a 5-10% boost since we avoid thrashing the L1-L3 cache by avoiding the memcopy to the page cache.
* However, `O_DIRECT` only bypasses the kernel's cache, there's still a cache beneath this in the disk itself. `O_DIRECT` only gets the data to the disk cache, and does nothing to ensure durability.

`O_DSYNC` means we toggle a bit when we send the write to the disk, so the disk knows to start writing right away and to only return when the write is durable. If we know that we always need durable writes, then we can use `O_DSYNC` to skip an `fsync()` call.